src/NHibernate/Async/Collection/Generic/PersistentGenericList.cs

```diff
@@ -0,0 +1,108 @@
+using NHibernate.Persister.Collection;
+using NHibernate.Type;
+using NHibernate.Util;
+
+namespace NHibernate.Collection.Generic
+{
+	using System.Threading.Tasks;
+	using System.Threading;
+	/// <content>
+	/// Contains generated async methods
+	/// </content>
+	public partial class PersistentGenericList<T> : AbstractPersistentCollection, IList<T>, IList
+	{
+
+		public override Task<ICollection> GetOrphansAsync(object snapshot, string entityName, CancellationToken cancellationToken)
+		{
+			if (cancellationToken.IsCancellationRequested)
+			{
+				return Task.FromCanceled<ICollection>(cancellationToken);
+			}
+			try
+			{
+				var sn = (IList<T>)snapshot;
+				return GetOrphansAsync((ICollection)sn, (ICollection) WrappedList, entityName, Session, cancellationToken);
+			}
+			catch (Exception ex)
+			{
+				return Task.FromException<ICollection>(ex);
+			}
+		}
+
+		public override async Task<object> ReadFromAsync(DbDataReader rs, ICollectionPersister role, ICollectionAliases descriptor, object owner, CancellationToken cancellationToken)
+		{
+			cancellationToken.ThrowIfCancellationRequested();
+			var element = (T)await (role.ReadElementAsync(rs, owner, descriptor.SuffixedElementAliases, Session, cancellationToken)).ConfigureAwait(false);
+			int index = (int)await (role.ReadIndexAsync(rs, descriptor.SuffixedIndexAliases, Session, cancellationToken)).ConfigureAwait(false);
+
+			//pad with nulls from the current last element up to the new index
+			for (int i = WrappedList.Count; i <= index; i++)
+			{
+				WrappedList.Insert(i, DefaultForType);
+			}
+
+			WrappedList[index] = element;
+			return element;
+		}
+
+		/// <summary>
+		/// Initializes this PersistentGenericList from the cached values.
+		/// </summary>
+		/// <param name="persister">The CollectionPersister to use to reassemble the PersistentGenericList.</param>
+		/// <param name="disassembled">The disassembled PersistentList.</param>
+		/// <param name="owner">The owner object.</param>
+		/// <param name="cancellationToken">A cancellation token that can be used to cancel the work</param>
+		public override async Task InitializeFromCacheAsync(ICollectionPersister persister, object disassembled, object owner, CancellationToken cancellationToken)
+		{
+			cancellationToken.ThrowIfCancellationRequested();
+			object[] array = (object[])disassembled;
+			int size = array.Length;
+			BeforeInitialize(persister, size);
+			for (int i = 0; i < size; i++)
+			{
+				var element = await (persister.ElementType.AssembleAsync(array[i], Session, owner, cancellationToken)).ConfigureAwait(false);
+				WrappedList.Add((T) (element ?? DefaultForType));
+			}
+		}
+
+		#region IList Members
+
+		#endregion
+		#region IList<T> Members
+
+		#endregion
+		#region ICollection Members
+
+		#endregion
+		#region ICollection<T> Members
+
+		#endregion
+		#region IEnumerable Members
+
+		#endregion
+		#region IEnumerable<T> Members
+
+		#endregion
+		#region DelayedOperations
+
+		#endregion
```

Comment:
[Added](https://github.com/maca88/AsyncGenerator/commit/7cbca5e3451934211624325a8ab53f8b335f3d22) a built-in plugin for removing empty regions.